### PR TITLE
[QoL] New Starter Select UI with Filter

### PR DESCRIPTION
**Pull request by KimJeongSun**
## What are the changes?

Add filter in starter select UI.
Most updates were cherry-picked from the previous [PR](https://github.com/pagefaultgames/pokerogue/pull/2036) with @FredeX's consent. 
This includes some bug fixes, test code updates, and minor code cleanups.

## Why am I doing these changes?

current starter UI is too difficult to find or check the over 500 pokemon.

## What did change?

added basic filters and merged the existing generation UI into the filter.

### Screenshots/Videos

https://github.com/user-attachments/assets/c5dbfedd-13f9-4318-a588-4f6cf913e232

This video has been updated based on this commit version. [8b64de5](https://github.com/pagefaultgames/pokerogue/pull/2916/commits/8b64de5119c7b54ba516ee701205ba697185387d)(latest)

## How to test the changes?

can test using the updated test code matching the updated UI.
`npm run test ui/starter-select`

## Checklist
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

### CODE REVIEW

**Comment by whatanoob** [comment]
Starter selection has been getting annoying lately, so this is very much needed.

It looks great but the one thing I'd also want would be the option to sort by # of candies. Although I haven't actually had time to go through the whole implementation, so I don't know how much work it would be to add that.

**Comment by FredeX** [comment]
Thank you for continuing my PR! can't wait to see it in game.

**Comment by Adri1** [comment]
It is possible to make it compatible with localization files?
And if possible, to have one translation entry per "All" option in the scrolling menus, as in some languages it's not the same word depentening on the contentent of a given menu
image

**Review by Adri1 — CHANGES_REQUESTED** [review]
It is possible to make it compatible with localization files?
And if possible, to have one translation entry per "All" option in the scrolling menus, as in some languages it's not the same word depentening on the contentent of a given menu
image

**Comment by KimJeongSun** [comment]
FYI. @flx-sta

As you mentioned, I have created a new branch to merge into the beta branch and opened a PR. I have tried to ensure there are no conflicts with the latest updates. I would appreciate it if you could review it. Thank you!! 🙇

**Comment by KimJeongSun** [comment]
@f-fsantos 
Is there a specific reason to set the merge target of this PR to the `beta` branch instead `main`? 

As I mentioned in the previous comment, there's already a [PR](https://github.com/pagefaultgames/pokerogue/pull/2950) branch based on the `beta` branch open for the `beta` branch merge. 

This PR is based on the main branch for merging into the `main` branch.

**Comment by KimJeongSun** [comment]
@f-fsantos  @flx-sta 
Upon inspecting with `git graph`, it appears that the `beta` branch is now fully prepared to function as the stage branch.

Please verify the following points:
1. Moving forward, the `beta` branch will operate as the stage, hence all features should be deployed in the order: beta -> main. Consequently, the target of this PR has been changed to the `beta` branch. Is this correct?
2. Considering the major overhaul done to the `beta` branch, it seems that continuing with the current `for beta` PR #2950  would no longer be meaningful. Therefore, I intend to close it.
3. I will incorporate the necessary bug fixes from the `for beta` PR into the current PR.

**Comment by KimJeongSun** [comment]
New Starter Select UI with Filter PR is fully prepared and ready to be merged into the beta branch! 🎉 

Based on the messages from managers yesterday, it seems that a significant force push was made to the beta branch to perform stage functions. 🙏  
As a result, the previously opened [PR](https://github.com/pagefaultgames/pokerogue/pull/2950) (which has now lost its purpose 😢 ) for merging beta branch has been closed, but additional updates for merging into the new beta branch have been made in this PR.

Please test it and provide your feedback! 🙇 

Also, if the beta branch functions well as a stage, it would be great if I could avoid creating new PRs for the main branch in the future. We've already gone through three rebase processes, and it's been quite painful. 🤯

**Comment by KimJeongSun** [comment]
FYI.
test video has been updated based on latest version.

**Comment by brain-frog** [comment]
I'd prefer if there was some way to see if each filter has been changed from their default so you can tell what filters are in use

**Comment by KimJeongSun** [comment]
> I'd prefer if there was some way to see if each filter has been changed from their default so you can tell what filters are in use

Thank you for your feedback.
How about changing color of filter menu font color to Green, if it is changed from default?

I will try to make some screenshots for these change. if you give confirmation, i will update in this PR

**Comment by KimJeongSun** [comment]
how about this? @brain-frog 

Set the default options as follows:
- Gen: All
- Type: All
- Unlocks: All
- Win: All
- Sort: No. ascending.

Whenever there is a difference from the default options, indicate that the filter is activated by displaying it in yellow.

I test this filter situation, I have set the options to:
- Gen: All,
- Type: Fire type,
- Unlocks: Shiny3/2/1,
- Win: All,
- Sort: Cost Descending.

During this time, the filters that differ from the default options will be displayed in yellow text as following screenshot.
image

**Comment by whatanoob** [comment]
There seems to be an issue with double inputs specifically with my controller (I use this controller for pokerogue and never ran into this). Oddly enough it works just fine when I use the keyboard.

<details><summary>I made a little video to demonstrate it.</summary>
<p>


https://github.com/user-attachments/assets/1874dabf-8a25-4285-8bd3-4544b38153ab



</p>
</details> 

It seems to only occur when starting from default values. Once I change one of the tabs from its default values, then I don't experience the double clicking anymore.

Runing latest commit ([fe55d5c](https://github.com/pagefaultgames/pokerogue/pull/2916/commits/fe55d5cd3104af5e9fbe5436abeccab8145add60))

File: src/ui/starter-select-ui-handler.ts [hunk 1]
```diff
@@ -309,13 +303,92 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
...
+    shiny2Sprite.setTint(getVariantTint(1));
+    const shiny3Sprite = this.scene.add.sprite(0, 0, "shiny_star_small");
+    shiny3Sprite.setTint(getVariantTint(2));
+
+    const unlocksOptions = [
+      new DropDownOption(this.scene, "SHINY3", null, shiny3Sprite),
+      new DropDownOption(this.scene, "SHINY2", null, shiny2Sprite),
+      new DropDownOption(this.scene, "SHINY", null, shiny1Sprite),
+      new DropDownOption(this.scene, "NORMAL", "Normal"),
+      new DropDownOption(this.scene, "UNCAUGHT", "Not Caught")];
```
Comment (whatanoob) [inline src/ui/starter-select-ui-handler.ts:369]:
```suggestion
      new DropDownOption(this.scene, "UNCAUGHT", "Not Caught"),
      new DropDownOption(this.scene, "PASSIVE", "Passive"),];
```

File: src/ui/starter-select-ui-handler.ts [hunk 2]
```diff
@@ -1800,132 +1851,227 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
...
+        // First, ensure you have the caught attributes for the species else default to bigint 0
+        const caughtVariants = this.scene.gameData.dexData[container.species.speciesId]?.caughtAttr || BigInt(0);
+
+        // Define the variables based on whether their respective variants have been caught
+        const isVariant3Caught = !!(caughtVariants & DexAttr.VARIANT_3);
+        const isVariant2Caught = !!(caughtVariants & DexAttr.VARIANT_2);
+        const isVariantCaught = !!(caughtVariants & DexAttr.SHINY);
+        const isCaught = !!(caughtVariants & DexAttr.NON_SHINY);
+        const isUncaught = !isCaught && !isVariantCaught && !isVariant2Caught && !isVariant3Caught;
+
```
Comment (whatanoob) [inline src/ui/starter-select-ui-handler.ts:1878]:
```suggestion
        const isPassiveUnlocked = this.scene.gameData.starterData[container.species.speciesId].passiveAttr > 0;

```

File: src/ui/starter-select-ui-handler.ts [hunk 3]
```diff
@@ -1800,132 +1851,227 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
...
+            return isVariant3Caught;
+          } else if (variant === "SHINY2") {
+            return isVariant2Caught && !isVariant3Caught;
+          } else if (variant === "SHINY") {
+            return isVariantCaught && !isVariant2Caught && !isVariant3Caught;
+          } else if (variant === "NORMAL") {
+            return isCaught && !isVariantCaught && !isVariant2Caught && !isVariant3Caught;
+          } else if (variant === "UNCAUGHT") {
+            return isUncaught;
+          }
```
Comment (whatanoob) [inline src/ui/starter-select-ui-handler.ts:1892]:
```suggestion
          } else if (variant === "PASSIVE") {
            return isPassiveUnlocked;
          }
```

**Review by whatanoob — CHANGES_REQUESTED** [review]
You might want to move it to a different position in the dropdown, but filtering passives unlocked seems to work as well.

Note that this is just a quick and dirty implementation that doesn't allow filtering for locked passives as well, which would probably be a good idea, but I really just wanted to pass on this suggestion made by some people.

https://github.com/user-attachments/assets/dc53d583-3904-429e-b55b-0b5a3f7b44f0

**Comment by KimJeongSun** [comment]
FYI @whatanoob 
Thank you for the review and comments. The new passive filter code suggestion also looks good!

However, the current PR already includes significant code changes, and adding more changes might complicate testing and merging. Nevertheless, since the passive filter is part of the further work coverage for the unlocks menu, and it doesn't seem to require a very complex function, I think we can add this feature.

Given the current design where each filter dropdown works with an "or" condition, it seems better to add both 'passive unlocked' and 'passive locked' options. Do you agree with this approach? (Of course, ensuring that these conditions are mutually exclusive.)

Regarding the issue of doubled input on the pad that was previously mentioned, my testing suggests that the problem may not originate from pad input itself but rather from a lag in processing the filter dropdown, leading to malfunction. While continuous input when holding down a button is a common approach across the entire pokerougue UI, the second input delay seems particularly short when toggling the 'all' option in the gen filter dropdown. I'll conduct further testing on the filter dropdown processing to diagnose and fix this bug (hopefully by today!).

File: src/ui/starter-select-ui-handler.ts [hunk 4]
```diff
@@ -2514,12 +2607,14 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
       textStyle = TextStyle.SUMMARY_GOLD;
       break;
     }
-    this.starterValueLabels[cursor].setColor(this.getTextColor(textStyle));
-    this.starterValueLabels[cursor].setShadowColor(this.getTextColor(textStyle, true));
+    if (baseStarterValue - starterValue > 0) {
+      starter.label.setColor(this.getTextColor(textStyle));
+      starter.label.setShadowColor(this.getTextColor(textStyle, true));
+    }
```
Comment (KimJeongSun) [inline src/ui/starter-select-ui-handler.ts:2613]: I analyzed the code to investigate the UI latency and discovered the bottleneck, which I have now fixed.

As a result, the UI update latency has reduced by up to 40ms, and the double input bug you mentioned has also been resolved. When I was debugging alone previously, I noticed occasional lags during scrolling caused by the same issue, which now seems to be fixed with this update. 🎉 

Thank you for the excellent bug report! 🙇  @whatanoob

Also, as always, I struggled with the merge conflicts of updated passive background feature.(please don't misunderstand! This feature is very convenient and cool!) 
I hope this branch gets merged into the beta soon so I can take a break from hobby development for a while! 😅 
FYI. @f-fsantos @flx-sta @brain-frog

Comment (whatanoob) [reply src/ui/starter-select-ui-handler.ts:2613]: Please note that I'm not actually a programmer - I just play the game and studied a bit of coding - so I can't do an actual code review 😁

But I can confirm that the double input is fixed now. And wow the navigation is so much smoother now... great job!

I also love the idea of highlighting passives unlocked.  For example, I can now see which of my shinies have them unlocked, while maintaining the "or" logic of the filters, which is what I reckon most people would find most intuitive. So, as an end user, I find that to be a really elegant solution, although some people might want it to be a toggleable feature.

I think now the only option missing is sorting by # of candies. I did a quick and dirty implementation and it works perfectly, except for starters that are not the "baby" in their evolutionary lines (e.g. Marill, Pikachu, Chimecho, etc.). That might be an easy fix, I just haven't looked into it yet.


<details><summary>Video</summary>
<p>

https://github.com/user-attachments/assets/509fd7aa-ff2e-4972-9698-0e4154b9f5d7

Changes:

starter-select-ui-handler.ts (line 376)
```
      new DropDownOption(this.scene, 3, "Name", null, DropDownState.OFF),
      new DropDownOption(this.scene, 4, "Candy", null, DropDownState.OFF),];
```

starter-select-ui-handler.ts (line 1911)
```
      case 4:
        const candyCountA = this.scene.gameData.starterData[a.species.speciesId].candyCount;
        const candyCountB = this.scene.gameData.starterData[b.species.speciesId].candyCount;
        return (candyCountA - candyCountB) * -sort.dir;
```

</p>
</details>

**Comment by KimJeongSun** [comment]
In fact, the update adding background colors to the passive unlock Pokémon backgrounds was carried out in this #2497 by @Fontbane , and we can look forward to making that an option in the next PR. 🙏 

As I mentioned, I don't wish to add many more updates, but since the sorting option you requested is easy to implement and test,  I added it. 👌 

Moving forward, I expect the filter UI to include various improvements, such as considering type changes for traits, moves, egg moves, and challenges in the future! 🚀

**Review by whatanoob — APPROVED** [review]

**Comment by torranx** [comment]
Did you test this on challenge mode?

**Comment by KimJeongSun** [comment]
> Did you test this on challenge mode?

I did. do you need test video that too?

**Comment by KimJeongSun** [comment]
This PR seems to have almost achieved its goal of adding a initial filter to the starter select UI.

I'm not entirely familiar with the pokerogue PR merge policy either, but based on what I've seen on Discord and from other PR updates, it seems that PRs like this one that bring changes to the UI require reviews from lead developers like you @brain-frog @Madmadness65 

If you don't mind, could you please review it? If you're too busy or unable to review it for any other reason, might I suggest delegating the review task to another developer? Thank you.

**Comment by KimJeongSun** [comment]
Hi @brain-frog @Madmadness65 , just a kind reminder for this PR review when you have a moment. Thanks!

**Comment by Tempo-anon** [comment]
A different PR was merged in that caused merge conflicts. Can you please fix those? Sorry about that

**Comment by KimJeongSun** [comment]
> A different PR was merged in that caused merge conflicts. Can you please fix those? Sorry about that

@Tempo-anon 
It's all good! This isn't anyone's fault; it's all part of the process of making a better game.
I've resolved the conflicts once more and ensured compatibility with the latest beta features to the best of my ability!

However, I'd really like to get this PR reviewed and wrapped up so I can move on to the next one!
FYI. @brain-frog

**Review by Madmadness65 — COMMENTED** [review]
The changed and added graphics need to be reflected onto the legacy UI as well. No additional code needed, just equivalent graphics in the `/ui/legacy/` folder.
image

**Comment by KimJeongSun** [comment]
First of all, thank you so much for the review!! I've been worried for a week now, wondering if I did something wrong. 😭 

As per your instructions, I have placed the missing graphics into the legacy folder. 
However, when checking the legacy UI, the added scroll bar looks quite ugly... 

Screenshot 2024-07-24 at 10 13 30 AM

so I have made some adjustments to the position and size of the starterContainerWindow for the legacy UI (this does not affect the existing UI, so no worries!). 
Although I have zero sense of art, I tried my best to place it in what I thought was the best position. 

Screenshot 2024-07-24 at 10 43 51 AM


Could you please review it again when you have a chance? @Madmadness65

**Review by Madmadness65 — APPROVED** [review]
After mixing and matching various filters, sorting options, and selecting Pokémon for runs with nothing breaking or visually missing, everything *seems* to be working as intended, so this LGTM. Sorry these reviews took so long, btw.

**Comment by KimJeongSun** [comment]
After seeing this comment, all my fatigue disappeared. 
Although I don't fully understand how busy you are as the lead dev, I'm cheering for you and look forward to working with you in the future. 👍

**Comment by KimJeongSun** [comment]
> After mixing and matching various filters, sorting options, and selecting Pokémon for runs with nothing breaking or visually missing, everything _seems_ to be working as intended, so this LGTM. Sorry these reviews took so long, btw.

After seeing this comment, all my fatigue disappeared.
Although I don't fully understand how busy you are as the lead dev, I'm cheering for you and look forward to working with you in the future. 👍

**Comment by flx-sta** [comment]
@KimJeongSun i found a little NIT:
- open “Gen” Drop down
- Select `All`
- Unselect Gen II
- `All` stays selected when it shouldn’t

**Comment by Opaque02** [comment]
The issue that flx mentioned above happens with every drop down, not just gen

**Comment by KimJeongSun** [comment]
i will make hotfix PR for this bugs in this[ discord thread](https://discord.com/channels/1125469663833370665/1265591426851409973/1265592525180108832) after i covered it
thanks @Opaque02 @flx-sta

**Pull request merged**